File: android/src/com/boliao/eod/NetworkRequestQueue.java
package com.boliao.eod;

import android.content.Context;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.Volley;

/**
 * A pseudo-singleton
 * - only one class init with context, the rest use singleton
 */
public class NetworkRequestQueue {
    private static NetworkRequestQueue instance;
    private RequestQueue requestQueue;
    private static Context context;

    static synchronized NetworkRequestQueue getInstance() {
//        if (instance == null)
//            instance = new NetworkRequestQueue(context);
        return instance;
    }

    static void init(Context context) {
        if (instance == null)
            instance = new NetworkRequestQueue(context);
    }

    private NetworkRequestQueue(Context context) {
        this.context = context;
        requestQueue = getRequestQueue();
    }

    RequestQueue getRequestQueue() {
        if (requestQueue == null) {
            // need to use getApplicationContext to prevent leaking activities/broadcastreceivers
            requestQueue = Volley.newRequestQueue(context.getApplicationContext());
        }
        return requestQueue;
    }

    <T> void add(Request<T> request) {
        getRequestQueue().add(request);
    }
}

File: android/src/com/boliao/eod/NetworkRequestQueue.kt
package com.boliao.eod

import android.content.Context
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.toolbox.Volley

/**
 * A singleton to handle requests
 * - allows someone to init object with context
 */
class NetworkRequestQueue constructor(context: Context) {
    companion object {
        private val TAG = NetworkRequestQueue::class.simpleName

        // Getting the singleton instance
        @Volatile
        private var INSTANCE: NetworkRequestQueue? = null
        fun getInstance(context: Context) = INSTANCE ?: synchronized(this) {
            INSTANCE ?: NetworkRequestQueue(context).also {
                INSTANCE = it
            }
        }
    }

    /**
     * A Volley RequestQueue to queue up requests
     * - use lazy init (by first person to use singleton)
     */
    val requestQueue: RequestQueue by lazy {
        Volley.newRequestQueue(context.applicationContext)
    }

    /**
     * Add Request of any type T into the request queue
     */
    fun <T> add(request: Request<T>?) {
        requestQueue.add(request)
    }
}